Test failures on Fedora Rawhide

##### SUMMARY
Attempting to build the collection on Fedora Rawhide we are seeing test failures.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
test_backend_openssl_cli.py
test_time.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible [core 2.16.12]
  config file = /home/orion/.ansible.cfg
  configured module search path = ['/home/orion/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.13/site-packages/ansible
  ansible collection location = /home/orion/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.13.0 (main, Oct  8 2024, 00:00:00) [GCC 14.2.1 20240912 (Red Hat 14.2.1-4)] (/usr/bin/python3)
  jinja version = 3.1.4
  libyaml = True
```

##### COLLECTION VERSION
<!--- Paste verbatim output from "ansible-galaxy collection list <namespace>.<collection>"  between the quotes
for example: ansible-galaxy collection list community.general
-->
```paste below
2.22.2
```

##### CONFIGURATION
<!--- Paste verbatim output from "ansible-config dump --only-changed" between quotes -->
```paste below
CONFIG_FILE() = /home/orion/.ansible.cfg
EDITOR(env: EDITOR) = vim
GALAXY_SERVER_LIST(/home/orion/.ansible.cfg) = ['release_galaxy']
PAGER(env: PAGER) = less
```

##### OS / ENVIRONMENT
Fedora Rawhide

##### STEPS TO REPRODUCE
<!--- Describe exactly how to reproduce the problem, using a minimal test-case -->

<!--- Paste example playbooks or commands between quotes below -->
```yaml
/usr/lib/rpm/ansible_collection.py test --python-interpreter /usr/bin/python3 --local
```

<!--- HINT: You can paste gist.github.com links for larger files -->

##### EXPECTED RESULTS
No test failures

##### ACTUAL RESULTS
```paste below
Loading collection metadata from /home/orion/fedora/ansible-collection-community-crypto/ansible-collection-community-crypto-2.22.2-build/community.crypto-2.22.2/galaxy.yml
Running: ('ansible-test', 'units', '--python-interpreter', '/usr/bin/python3', '--local')

Unit test modules with Python 3.13
============================= test session starts ==============================
platform linux -- Python 3.13.0, pytest-8.3.3, pluggy-1.5.0
rootdir: /tmp/tmpyv_e91xd/ansible_collections/community/crypto
configfile: ../../../../../usr/lib/python3.13/site-packages/ansible_test/_data/pytest/config/default.ini
plugins: hypothesis-6.104.2, libtmux-0.37.0, flaky-3.8.1, asyncio-0.23.6, rerunfailures-14.0, forked-1.6.0, requests-mock-1.12.1, anyio-3.7.1, openfiles-0.6.0, pyfakefs-5.6.0, vcr-1.0.2, arraydiff-0.6.1, xprocess-1.0.2, mock-3.14.0, remotedata-0.4.1, mpi-0.6, xdist-3.6.1, timeout-2.3.1, doctestplus-1.2.1, localserver-0.9.0.post0
asyncio: mode=Mode.STRICT
created: 6/6 workers
6 workers [76 items]

........................................................................ [ 94%]
....                                                                     [100%]
- generated xml file: /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/output/junit/python3.13-modules-units.xml -
============================== 76 passed in 1.91s ==============================
Unit test module_utils with Python 3.13
============================= test session starts ==============================
platform linux -- Python 3.13.0, pytest-8.3.3, pluggy-1.5.0
rootdir: /tmp/tmpyv_e91xd/ansible_collections/community/crypto
configfile: ../../../../../usr/lib/python3.13/site-packages/ansible_test/_data/pytest/config/default.ini
plugins: hypothesis-6.104.2, libtmux-0.37.0, flaky-3.8.1, asyncio-0.23.6, rerunfailures-14.0, forked-1.6.0, requests-mock-1.12.1, anyio-3.7.1, openfiles-0.6.0, pyfakefs-5.6.0, vcr-1.0.2, arraydiff-0.6.1, xprocess-1.0.2, mock-3.14.0, remotedata-0.4.1, mpi-0.6, xdist-3.6.1, timeout-2.3.1, doctestplus-1.2.1, localserver-0.9.0.post0
asyncio: mode=Mode.STRICT
created: 6/6 workers
6 workers [339 items]

................F.................F..................................... [ 21%]
...s.....s....s...ss...s..s.ss...s....F..s.....s....s......s.......s.... [ 42%]
........................................................................ [ 63%]
........................................................................ [ 84%]
.........FFF...F.F.................................                      [100%]
=================================== FAILURES ===================================
____________ test_interpolate_timestamp[start1-end1-0.5-expected1] _____________
[gw3] linux -- Python 3.13.0 /usr/bin/python3

start = {'day': 1, 'hour': 0, 'minute': 0, 'month': 1, ...}
end = {'day': 1, 'hour': 1, 'minute': 0, 'month': 1, ...}, percentage = 0.5
expected = {'day': 1, 'hour': 0, 'minute': 30, 'month': 1, ...}

    @pytest.mark.parametrize("start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
    def test_interpolate_timestamp(start, end, percentage, expected):
        module = MagicMock()
        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
        ts_start = backend.get_utc_datetime(**start)
        ts_end = backend.get_utc_datetime(**end)
        ts_expected = backend.get_utc_datetime(**expected)
        timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
>       assert ts_expected == timestamp
E       assert datetime.datetime(2024, 1, 1, 0, 30) == datetime.datetime(2024, 1, 1, 7, 30)

tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py:122: AssertionError
____________ test_interpolate_timestamp[start2-end2-1.0-expected2] _____________
[gw3] linux -- Python 3.13.0 /usr/bin/python3

start = {'day': 1, 'hour': 0, 'minute': 0, 'month': 1, ...}
end = {'day': 1, 'hour': 1, 'minute': 0, 'month': 1, ...}, percentage = 1.0
expected = {'day': 1, 'hour': 1, 'minute': 0, 'month': 1, ...}

    @pytest.mark.parametrize("start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
    def test_interpolate_timestamp(start, end, percentage, expected):
        module = MagicMock()
        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
        ts_start = backend.get_utc_datetime(**start)
        ts_end = backend.get_utc_datetime(**end)
        ts_expected = backend.get_utc_datetime(**expected)
        timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
>       assert ts_expected == timestamp
E       assert datetime.datetime(2024, 1, 1, 1, 0) == datetime.datetime(2024, 1, 1, 8, 0)

tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py:122: AssertionError
____________ test_interpolate_timestamp[start0-end0-0.0-expected0] _____________
[gw2] linux -- Python 3.13.0 /usr/bin/python3

start = {'day': 1, 'hour': 0, 'minute': 0, 'month': 1, ...}
end = {'day': 1, 'hour': 1, 'minute': 0, 'month': 1, ...}, percentage = 0.0
expected = {'day': 1, 'hour': 0, 'minute': 0, 'month': 1, ...}

    @pytest.mark.parametrize("start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
    def test_interpolate_timestamp(start, end, percentage, expected):
        module = MagicMock()
        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
        ts_start = backend.get_utc_datetime(**start)
        ts_end = backend.get_utc_datetime(**end)
        ts_expected = backend.get_utc_datetime(**expected)
        timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
>       assert ts_expected == timestamp
E       assert datetime.datetime(2024, 1, 1, 0, 0) == datetime.datetime(2024, 1, 1, 7, 0)

tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py:122: AssertionError
_______________________ test_epoch_seconds[0-timestamp0] _______________________
[gw2] linux -- Python 3.13.0 /usr/bin/python3

seconds = 0
timestamp = {'day': 1, 'hour': 0, 'microsecond': 0, 'minute': 0, ...}

    @pytest.mark.parametrize("seconds, timestamp", TEST_EPOCH_SECONDS)
    def test_epoch_seconds(seconds, timestamp):
        ts_wo_tz = datetime.datetime(**timestamp)
>       assert seconds == get_epoch_seconds(ts_wo_tz)
E       assert 0 == 25200.0
E        +  where 25200.0 = get_epoch_seconds(datetime.datetime(1970, 1, 1, 0, 0))

tests/unit/plugins/module_utils/test_time.py:298: AssertionError
_____________________ test_epoch_seconds[1e-06-timestamp1] _____________________
[gw4] linux -- Python 3.13.0 /usr/bin/python3

seconds = 1e-06
timestamp = {'day': 1, 'hour': 0, 'microsecond': 1, 'minute': 0, ...}

    @pytest.mark.parametrize("seconds, timestamp", TEST_EPOCH_SECONDS)
    def test_epoch_seconds(seconds, timestamp):
        ts_wo_tz = datetime.datetime(**timestamp)
>       assert seconds == get_epoch_seconds(ts_wo_tz)
E       assert 1e-06 == 25200.000001
E        +  where 25200.000001 = get_epoch_seconds(datetime.datetime(1970, 1, 1, 0, 0, 0, 1))

tests/unit/plugins/module_utils/test_time.py:298: AssertionError
_____________________ test_epoch_seconds[0.001-timestamp2] _____________________
[gw0] linux -- Python 3.13.0 /usr/bin/python3

seconds = 0.001
timestamp = {'day': 1, 'hour': 0, 'microsecond': 1000, 'minute': 0, ...}

    @pytest.mark.parametrize("seconds, timestamp", TEST_EPOCH_SECONDS)
    def test_epoch_seconds(seconds, timestamp):
        ts_wo_tz = datetime.datetime(**timestamp)
>       assert seconds == get_epoch_seconds(ts_wo_tz)
E       assert 0.001 == 25200.001
E        +  where 25200.001 = get_epoch_seconds(datetime.datetime(1970, 1, 1, 0, 0, 0, 1000))

tests/unit/plugins/module_utils/test_time.py:298: AssertionError
____________________ test_epoch_seconds[3691.2-timestamp3] _____________________
[gw2] linux -- Python 3.13.0 /usr/bin/python3

seconds = 3691.2
timestamp = {'day': 1, 'hour': 1, 'microsecond': 200000, 'minute': 1, ...}

    @pytest.mark.parametrize("seconds, timestamp", TEST_EPOCH_SECONDS)
    def test_epoch_seconds(seconds, timestamp):
        ts_wo_tz = datetime.datetime(**timestamp)
>       assert seconds == get_epoch_seconds(ts_wo_tz)
E       assert 3691.2 == 28891.2
E        +  where 28891.2 = get_epoch_seconds(datetime.datetime(1970, 1, 1, 1, 1, 31, 200000))

tests/unit/plugins/module_utils/test_time.py:298: AssertionError
_____________________ test_epoch_to_seconds[timestamp0-62] _____________________
[gw4] linux -- Python 3.13.0 /usr/bin/python3

timestamp = datetime.datetime(1970, 1, 1, 0, 1, 2), expected_seconds = 62

    @pytest.mark.parametrize("timestamp, expected_seconds", TEST_EPOCH_TO_SECONDS)
    def test_epoch_to_seconds(timestamp, expected_seconds):
>       assert expected_seconds == get_epoch_seconds(timestamp)
E       assert 62 == 25262.0
E        +  where 25262.0 = get_epoch_seconds(datetime.datetime(1970, 1, 1, 0, 1, 2))

tests/unit/plugins/module_utils/test_time.py:311: AssertionError
=============================== warnings summary ===============================
tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start1-end1-0.5-expected1]
tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start2-end2-1.0-expected2]
tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start0-end0-0.0-expected0]
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/plugins/module_utils/time.py:92: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    return datetime.datetime.utcfromtimestamp(timestamp)

tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_now
tests/unit/plugins/module_utils/test_time.py::test_get_now_datetime
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/plugins/module_utils/time.py:49: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    return datetime.datetime.utcnow()

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_default_key_params[dsa-None-None-None]
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:151: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    pair = OpensshKeypair.generate(keytype=keytype, size=size, passphrase=passphrase, comment=comment)

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_default_key_params[dsa-None-None-None]
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:153: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    pair = OpensshKeypair.generate(keytype=keytype, size=size, passphrase=passphrase, comment=comment)

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:351: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    assert OpensshKeypair.generate() != OpensshKeypair.generate(keytype='dsa')

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:363: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    'pair': OpensshKeypair.generate(keytype='dsa', passphrase='change_me'.encode('UTF-8')),

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/plugins/module_utils/openssh/cryptography.py:560: CryptographyDeprecationWarning: SSH DSA keys are deprecated and will be removed in a future release.
    privatekey = privatekey_loader(

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/plugins/module_utils/openssh/cryptography.py:615: CryptographyDeprecationWarning: SSH DSA keys are deprecated and will be removed in a future release.
    publickey = publickey_loader(

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:380: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    loaded_dsa_key = OpensshKeypair.load(path=keys['dsa']['filename'], passphrase='change_me'.encode('UTF-8'))

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:383: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    loaded_dsa_key.update_passphrase('change_me_again'.encode('UTF-8'))

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_keypair_comparison
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:386: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    loaded_dsa_key.update_passphrase('change_me'.encode('UTF-8'))

tests/unit/plugins/module_utils/openssh/test_cryptography.py::test_valid_user_key_params[dsa-1024-change_me-comment]
  /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/unit/plugins/module_utils/openssh/test_cryptography.py:169: CryptographyDeprecationWarning: SSH DSA key support is deprecated and will be removed in a future release
    pair = OpensshKeypair.generate(keytype=keytype, size=size, passphrase=passphrase, comment=comment)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
- generated xml file: /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/output/junit/python3.13-module_utils-units.xml -
=========================== short test summary info ============================
SKIPPED [15] tests/unit/plugins/module_utils/crypto/test_asn1.py:78: unconditional skip
FAILED tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start1-end1-0.5-expected1]
FAILED tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start2-end2-1.0-expected2]
FAILED tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py::test_interpolate_timestamp[start0-end0-0.0-expected0]
FAILED tests/unit/plugins/module_utils/test_time.py::test_epoch_seconds[0-timestamp0]
FAILED tests/unit/plugins/module_utils/test_time.py::test_epoch_seconds[1e-06-timestamp1]
FAILED tests/unit/plugins/module_utils/test_time.py::test_epoch_seconds[0.001-timestamp2]
FAILED tests/unit/plugins/module_utils/test_time.py::test_epoch_seconds[3691.2-timestamp3]
FAILED tests/unit/plugins/module_utils/test_time.py::test_epoch_to_seconds[timestamp0-62]
============ 8 failed, 316 passed, 15 skipped, 15 warnings in 9.80s ============
FATAL: Command "pytest -r a -n auto --color no -p no:cacheprovider -c /usr/lib/python3.13/site-packages/ansible_test/_data/pytest/config/default.ini --junit-xml /tmp/tmpyv_e91xd/ansible_collections/community/crypto/tests/output/junit/python3.13-module_utils-units.xml --strict-markers --rootdir /tmp/tmpyv_e91xd/ansible_collections/community/crypto --confcutdir /tmp/tmpyv_e91xd/ansible_collections/community/crypto tests/unit/plugins/module_utils/acme/test_backend_cryptography.py tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py tests/unit/plugins/module_utils/acme/test_challenges.py tests/unit/plugins/module_utils/acme/test_errors.py tests/unit/plugins/module_utils/acme/test_io.py tests/unit/plugins/module_utils/acme/test_orders.py tests/unit/plugins/module_utils/acme/test_utils.py tests/unit/plugins/module_utils/crypto/test_asn1.py tests/unit/plugins/module_utils/crypto/test_cryptography_support.py tests/unit/plugins/module_utils/crypto/test_math.py tests/unit/plugins/module_utils/crypto/test_pem.py tests/unit/plugins/module_utils/openssh/test_certificate.py tests/unit/plugins/module_utils/openssh/test_cryptography.py tests/unit/plugins/module_utils/openssh/test_utils.py tests/unit/plugins/module_utils/test_time.py" returned exit status 1.
Command '('ansible-test', 'units', '--python-interpreter', '/usr/bin/python3', '--local')' returned non-zero exit status 1.
```

--- a/changelogs/fragments/810-time.yml
+++ b/changelogs/fragments/810-time.yml
@@ -1,0 +1,2 @@
+bugfixes:
+  - "time module utils - fix conversion of naive ``datetime`` objects to UNIX timestamps for Python 3 (https://github.com/ansible-collections/community.crypto/issues/808, https://github.com/ansible-collections/community.crypto/pull/810)."

--- a/plugins/module_utils/time.py
+++ b/plugins/module_utils/time.py
@@ -83,6 +83,9 @@ if sys.version_info < (3, 3):
             return (delta.microseconds + (delta.seconds + delta.days * 24 * 3600) * 10**6) / 10**6
 else:
     def get_epoch_seconds(timestamp):
+        if timestamp.tzinfo is None:
+            # timestamp.timestamp() is offset by the local timezone if timestamp has no timezone
+            timestamp = ensure_utc_timezone(timestamp)
         return timestamp.timestamp()
 
 

--- a/tests/unit/plugins/module_utils/acme/backend_data.py
+++ b/tests/unit/plugins/module_utils/acme/backend_data.py
@@ -20,6 +20,8 @@ from ansible_collections.community.crypto.plugins.module_utils.acme.errors impor
     BackendException,
 )
 
+from ..test_time import cartesian_product, TIMEZONES
+
 
 def load_fixture(name):
     with open(os.path.join(os.path.dirname(__file__), 'fixtures', name)) as f:
@@ -89,11 +91,11 @@ TEST_CERT_OPENSSL_OUTPUT_2 = load_fixture("cert_2.txt")  # OpenSSL 3.3.0 output
 TEST_CERT_OPENSSL_OUTPUT_2B = load_fixture("cert_2-b.txt")  # OpenSSL 1.1.1f output
 
 
-TEST_CERT_DAYS = [
+TEST_CERT_DAYS = cartesian_product(TIMEZONES, [
     (datetime.datetime(2018, 11, 15, 1, 2, 3), 11),
     (datetime.datetime(2018, 11, 25, 15, 20, 0), 1),
     (datetime.datetime(2018, 11, 25, 15, 30, 0), 0),
-]
+])
 
 
 TEST_CERT_INFO = CertificateInformation(
@@ -121,7 +123,7 @@ TEST_CERT_INFO = [
 ]
 
 
-TEST_PARSE_ACME_TIMESTAMP = [
+TEST_PARSE_ACME_TIMESTAMP = cartesian_product(TIMEZONES, [
     (
         '2024-01-01T00:11:22Z',
         dict(year=2024, month=1, day=1, hour=0, minute=11, second=22),
@@ -134,10 +136,10 @@ TEST_PARSE_ACME_TIMESTAMP = [
         '2024-04-17T06:54:13.333333334Z',
         dict(year=2024, month=4, day=17, hour=6, minute=54, second=13, microsecond=333333),
     ),
-]
+])
 
 if sys.version_info >= (3, 5):
-    TEST_PARSE_ACME_TIMESTAMP.extend([
+    TEST_PARSE_ACME_TIMESTAMP.extend(cartesian_product(TIMEZONES, [
         (
             '2024-01-01T00:11:22+0100',
             dict(year=2023, month=12, day=31, hour=23, minute=11, second=22),
@@ -146,10 +148,10 @@ if sys.version_info >= (3, 5):
             '2024-01-01T00:11:22.123+0100',
             dict(year=2023, month=12, day=31, hour=23, minute=11, second=22, microsecond=123000),
         ),
-    ])
+    ]))
 
 
-TEST_INTERPOLATE_TIMESTAMP = [
+TEST_INTERPOLATE_TIMESTAMP = cartesian_product(TIMEZONES, [
     (
         dict(year=2024, month=1, day=1, hour=0, minute=0, second=0),
         dict(year=2024, month=1, day=1, hour=1, minute=0, second=0),
@@ -168,7 +170,7 @@ TEST_INTERPOLATE_TIMESTAMP = [
         1.0,
         dict(year=2024, month=1, day=1, hour=1, minute=0, second=0),
     ),
-]
+])
 
 
 class FakeBackend(CryptoBackend):

--- a/tests/unit/plugins/module_utils/acme/test_backend_cryptography.py
+++ b/tests/unit/plugins/module_utils/acme/test_backend_cryptography.py
@@ -5,11 +5,14 @@
 from __future__ import absolute_import, division, print_function
 __metaclass__ = type
 
+import datetime
 
 import pytest
+from freezegun import freeze_time
 
 from ansible_collections.community.crypto.tests.unit.compat.mock import MagicMock
 
+from ansible_collections.community.crypto.plugins.module_utils.time import UTC
 
 from ansible_collections.community.crypto.plugins.module_utils.acme.backend_cryptography import (
     HAS_CURRENT_CRYPTOGRAPHY,
@@ -33,6 +36,8 @@ from .backend_data import (
     TEST_PARSE_ACME_TIMESTAMP,
     TEST_INTERPOLATE_TIMESTAMP,
 )
+
+from ..test_time import TIMEZONES
 
 
 if not HAS_CURRENT_CRYPTOGRAPHY:
@@ -65,16 +70,17 @@ def test_csridentifiers_cryptography(csr, result, openssl_output, tmpdir):
     assert identifiers == result
 
 
-@pytest.mark.parametrize("now, expected_days", TEST_CERT_DAYS)
-def test_certdays_cryptography(now, expected_days, tmpdir):
-    fn = tmpdir / 'test-cert.pem'
-    fn.write(TEST_CERT)
-    module = MagicMock()
-    backend = CryptographyBackend(module)
-    days = backend.get_cert_days(cert_filename=str(fn), now=now)
-    assert days == expected_days
-    days = backend.get_cert_days(cert_content=TEST_CERT, now=now)
-    assert days == expected_days
+@pytest.mark.parametrize("timezone, now, expected_days", TEST_CERT_DAYS)
+def test_certdays_cryptography(timezone, now, expected_days, tmpdir):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        fn = tmpdir / 'test-cert.pem'
+        fn.write(TEST_CERT)
+        module = MagicMock()
+        backend = CryptographyBackend(module)
+        days = backend.get_cert_days(cert_filename=str(fn), now=now)
+        assert days == expected_days
+        days = backend.get_cert_days(cert_content=TEST_CERT, now=now)
+        assert days == expected_days
 
 
 @pytest.mark.parametrize("cert_content, expected_cert_info, openssl_output", TEST_CERT_INFO)
@@ -96,28 +102,40 @@ def test_get_cert_information(cert_content, expected_cert_info, openssl_output, 
     assert cert_info == expected_cert_info
 
 
-def test_now():
-    module = MagicMock()
-    backend = CryptographyBackend(module)
-    now = backend.get_now()
-    assert CRYPTOGRAPHY_TIMEZONE == (now.tzinfo is not None)
+# @pytest.mark.parametrize("timezone", TIMEZONES)
+# Due to a bug in freezegun (https://github.com/spulec/freezegun/issues/348, https://github.com/spulec/freezegun/issues/553)
+# this only works with timezone = UTC if CRYPTOGRAPHY_TIMEZONE is truish
+@pytest.mark.parametrize("timezone", [datetime.timedelta(hours=0)] if CRYPTOGRAPHY_TIMEZONE else TIMEZONES)
+def test_now(timezone):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        module = MagicMock()
+        backend = CryptographyBackend(module)
+        now = backend.get_now()
+        if CRYPTOGRAPHY_TIMEZONE:
+            assert now.tzinfo is not None
+            assert now == datetime.datetime(2024, 2, 3, 4, 5, 6, tzinfo=UTC)
+        else:
+            assert now.tzinfo is None
+            assert now == datetime.datetime(2024, 2, 3, 4, 5, 6)
 
 
-@pytest.mark.parametrize("input, expected", TEST_PARSE_ACME_TIMESTAMP)
-def test_parse_acme_timestamp(input, expected):
-    module = MagicMock()
-    backend = CryptographyBackend(module)
-    ts_expected = backend.get_utc_datetime(**expected)
-    timestamp = backend.parse_acme_timestamp(input)
-    assert ts_expected == timestamp
+@pytest.mark.parametrize("timezone, input, expected", TEST_PARSE_ACME_TIMESTAMP)
+def test_parse_acme_timestamp(timezone, input, expected):
+    with freeze_time("2024-02-03 04:05:06 +00:00", tz_offset=timezone):
+        module = MagicMock()
+        backend = CryptographyBackend(module)
+        ts_expected = backend.get_utc_datetime(**expected)
+        timestamp = backend.parse_acme_timestamp(input)
+        assert ts_expected == timestamp
 
 
-@pytest.mark.parametrize("start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
-def test_interpolate_timestamp(start, end, percentage, expected):
-    module = MagicMock()
-    backend = CryptographyBackend(module)
-    ts_start = backend.get_utc_datetime(**start)
-    ts_end = backend.get_utc_datetime(**end)
-    ts_expected = backend.get_utc_datetime(**expected)
-    timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
-    assert ts_expected == timestamp
+@pytest.mark.parametrize("timezone, start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
+def test_interpolate_timestamp(timezone, start, end, percentage, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        module = MagicMock()
+        backend = CryptographyBackend(module)
+        ts_start = backend.get_utc_datetime(**start)
+        ts_end = backend.get_utc_datetime(**end)
+        ts_expected = backend.get_utc_datetime(**expected)
+        timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
+        assert ts_expected == timestamp

--- a/tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py
+++ b/tests/unit/plugins/module_utils/acme/test_backend_openssl_cli.py
@@ -5,8 +5,10 @@
 from __future__ import absolute_import, division, print_function
 __metaclass__ = type
 
+import datetime
 
 import pytest
+from freezegun import freeze_time
 
 from ansible_collections.community.crypto.tests.unit.compat.mock import MagicMock
 
@@ -25,6 +27,8 @@ from .backend_data import (
     TEST_PARSE_ACME_TIMESTAMP,
     TEST_INTERPOLATE_TIMESTAMP,
 )
+
+from ..test_time import TIMEZONES
 
 
 TEST_IPS = [
@@ -69,17 +73,18 @@ def test_normalize_ip(ip, result):
     assert backend._normalize_ip(ip) == result
 
 
-@pytest.mark.parametrize("now, expected_days", TEST_CERT_DAYS)
-def test_certdays_cryptography(now, expected_days, tmpdir):
-    fn = tmpdir / 'test-cert.pem'
-    fn.write(TEST_CERT)
-    module = MagicMock()
-    module.run_command = MagicMock(return_value=(0, TEST_CERT_OPENSSL_OUTPUT, 0))
-    backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
-    days = backend.get_cert_days(cert_filename=str(fn), now=now)
-    assert days == expected_days
-    days = backend.get_cert_days(cert_content=TEST_CERT, now=now)
-    assert days == expected_days
+@pytest.mark.parametrize("timezone, now, expected_days", TEST_CERT_DAYS)
+def test_certdays_cryptography(timezone, now, expected_days, tmpdir):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        fn = tmpdir / 'test-cert.pem'
+        fn.write(TEST_CERT)
+        module = MagicMock()
+        module.run_command = MagicMock(return_value=(0, TEST_CERT_OPENSSL_OUTPUT, 0))
+        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
+        days = backend.get_cert_days(cert_filename=str(fn), now=now)
+        assert days == expected_days
+        days = backend.get_cert_days(cert_content=TEST_CERT, now=now)
+        assert days == expected_days
 
 
 @pytest.mark.parametrize("cert_content, expected_cert_info, openssl_output", TEST_CERT_INFO)
@@ -95,28 +100,33 @@ def test_get_cert_information(cert_content, expected_cert_info, openssl_output, 
     assert cert_info == expected_cert_info
 
 
-def test_now():
-    module = MagicMock()
-    backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
-    now = backend.get_now()
-    assert now.tzinfo is None
+@pytest.mark.parametrize("timezone", TIMEZONES)
+def test_now(timezone):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        module = MagicMock()
+        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
+        now = backend.get_now()
+        assert now.tzinfo is None
+        assert now == datetime.datetime(2024, 2, 3, 4, 5, 6)
 
 
-@pytest.mark.parametrize("input, expected", TEST_PARSE_ACME_TIMESTAMP)
-def test_parse_acme_timestamp(input, expected):
-    module = MagicMock()
-    backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
-    ts_expected = backend.get_utc_datetime(**expected)
-    timestamp = backend.parse_acme_timestamp(input)
-    assert ts_expected == timestamp
+@pytest.mark.parametrize("timezone, input, expected", TEST_PARSE_ACME_TIMESTAMP)
+def test_parse_acme_timestamp(timezone, input, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        module = MagicMock()
+        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
+        ts_expected = backend.get_utc_datetime(**expected)
+        timestamp = backend.parse_acme_timestamp(input)
+        assert ts_expected == timestamp
 
 
-@pytest.mark.parametrize("start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
-def test_interpolate_timestamp(start, end, percentage, expected):
-    module = MagicMock()
-    backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
-    ts_start = backend.get_utc_datetime(**start)
-    ts_end = backend.get_utc_datetime(**end)
-    ts_expected = backend.get_utc_datetime(**expected)
-    timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
-    assert ts_expected == timestamp
+@pytest.mark.parametrize("timezone, start, end, percentage, expected", TEST_INTERPOLATE_TIMESTAMP)
+def test_interpolate_timestamp(timezone, start, end, percentage, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        module = MagicMock()
+        backend = OpenSSLCLIBackend(module, openssl_binary='openssl')
+        ts_start = backend.get_utc_datetime(**start)
+        ts_end = backend.get_utc_datetime(**end)
+        ts_expected = backend.get_utc_datetime(**expected)
+        timestamp = backend.interpolate_timestamp(ts_start, ts_end, percentage)
+        assert ts_expected == timestamp

--- a/tests/unit/plugins/module_utils/test_time.py
+++ b/tests/unit/plugins/module_utils/test_time.py
@@ -10,7 +10,9 @@ import datetime
 import sys
 
 import pytest
+from freezegun import freeze_time
 
+from ansible.module_utils.common.collections import is_sequence
 
 from ansible_collections.community.crypto.plugins.module_utils.time import (
     add_or_remove_timezone,
@@ -25,18 +27,42 @@ from ansible_collections.community.crypto.plugins.module_utils.time import (
 )
 
 
-TEST_REMOVE_TIMEZONE = [
-    (
-        datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=UTC),
-        datetime.datetime(2024, 1, 1, 0, 1, 2),
-    ),
-    (
-        datetime.datetime(2024, 1, 1, 0, 1, 2),
-        datetime.datetime(2024, 1, 1, 0, 1, 2),
-    ),
+TIMEZONES = [
+    datetime.timedelta(hours=0),
+    datetime.timedelta(hours=1),
+    datetime.timedelta(hours=2),
+    datetime.timedelta(hours=-6),
 ]
 
-TEST_UTC_TIMEZONE = [
+
+def cartesian_product(list1, list2):
+    result = []
+    for item1 in list1:
+        if not is_sequence(item1):
+            item1 = (item1, )
+        elif not isinstance(item1, tuple):
+            item1 = tuple(item1)
+        for item2 in list2:
+            if not is_sequence(item2):
+                item2 = (item2, )
+            elif not isinstance(item2, tuple):
+                item2 = tuple(item2)
+            result.append(item1 + item2)
+    return result
+
+
+TEST_REMOVE_TIMEZONE = cartesian_product(TIMEZONES, [
+    (
+        datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=UTC),
+        datetime.datetime(2024, 1, 1, 0, 1, 2),
+    ),
+    (
+        datetime.datetime(2024, 1, 1, 0, 1, 2),
+        datetime.datetime(2024, 1, 1, 0, 1, 2),
+    ),
+])
+
+TEST_UTC_TIMEZONE = cartesian_product(TIMEZONES, [
     (
         datetime.datetime(2024, 1, 1, 0, 1, 2),
         datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=UTC),
@@ -45,21 +71,21 @@ TEST_UTC_TIMEZONE = [
         datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=UTC),
         datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=UTC),
     ),
-]
+])
 
-TEST_EPOCH_SECONDS = [
+TEST_EPOCH_SECONDS = cartesian_product(TIMEZONES, [
     (0, dict(year=1970, day=1, month=1, hour=0, minute=0, second=0, microsecond=0)),
     (1E-6, dict(year=1970, day=1, month=1, hour=0, minute=0, second=0, microsecond=1)),
     (1E-3, dict(year=1970, day=1, month=1, hour=0, minute=0, second=0, microsecond=1000)),
     (3691.2, dict(year=1970, day=1, month=1, hour=1, minute=1, second=31, microsecond=200000)),
-]
+])
 
-TEST_EPOCH_TO_SECONDS = [
+TEST_EPOCH_TO_SECONDS = cartesian_product(TIMEZONES, [
     (datetime.datetime(1970, 1, 1, 0, 1, 2, 0), 62),
     (datetime.datetime(1970, 1, 1, 0, 1, 2, 0, tzinfo=UTC), 62),
-]
+])
 
-TEST_CONVERT_RELATIVE_TO_DATETIME = [
+TEST_CONVERT_RELATIVE_TO_DATETIME = cartesian_product(TIMEZONES, [
     (
         '+0',
         False,
@@ -96,9 +122,9 @@ TEST_CONVERT_RELATIVE_TO_DATETIME = [
         datetime.datetime(2024, 1, 1, 0, 0, 0),
         datetime.datetime(2023, 10, 1, 17, 19, 10, tzinfo=UTC),
     ),
-]
+])
 
-TEST_GET_RELATIVE_TIME_OPTION = [
+TEST_GET_RELATIVE_TIME_OPTION = cartesian_product(TIMEZONES, [
     (
         '+1d2h3m4s',
         'foo',
@@ -195,28 +221,28 @@ TEST_GET_RELATIVE_TIME_OPTION = [
         datetime.datetime(2024, 1, 1, 0, 0, 0),
         '202401020405Z',
     ),
-]
+])
 
 
 if sys.version_info >= (3, 5):
     ONE_HOUR_PLUS = datetime.timezone(datetime.timedelta(hours=1))
 
-    TEST_REMOVE_TIMEZONE.extend([
+    TEST_REMOVE_TIMEZONE.extend(cartesian_product(TIMEZONES, [
         (
             datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=ONE_HOUR_PLUS),
             datetime.datetime(2023, 12, 31, 23, 1, 2),
         ),
-    ])
-    TEST_UTC_TIMEZONE.extend([
+    ]))
+    TEST_UTC_TIMEZONE.extend(cartesian_product(TIMEZONES, [
         (
             datetime.datetime(2024, 1, 1, 0, 1, 2, tzinfo=ONE_HOUR_PLUS),
             datetime.datetime(2023, 12, 31, 23, 1, 2, tzinfo=UTC),
         ),
-    ])
-    TEST_EPOCH_TO_SECONDS.extend([
+    ]))
+    TEST_EPOCH_TO_SECONDS.extend(cartesian_product(TIMEZONES, [
         (datetime.datetime(1970, 1, 1, 0, 1, 2, 0, tzinfo=ONE_HOUR_PLUS), 62 - 3600),
-    ])
-    TEST_GET_RELATIVE_TIME_OPTION.extend([
+    ]))
+    TEST_GET_RELATIVE_TIME_OPTION.extend(cartesian_product(TIMEZONES, [
         (
             '20240102040506+0100',
             'foo',
@@ -265,59 +291,77 @@ if sys.version_info >= (3, 5):
             datetime.datetime(2024, 1, 1, 0, 0, 0),
             '202401020405+0100',
         ),
-    ])
+    ]))
 
 
-@pytest.mark.parametrize("input, expected", TEST_REMOVE_TIMEZONE)
-def test_remove_timezone(input, expected):
-    output_1 = remove_timezone(input)
-    assert expected == output_1
-    output_2 = add_or_remove_timezone(input, with_timezone=False)
-    assert expected == output_2
+@pytest.mark.parametrize("timezone, input, expected", TEST_REMOVE_TIMEZONE)
+def test_remove_timezone(timezone, input, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output_1 = remove_timezone(input)
+        assert expected == output_1
+        output_2 = add_or_remove_timezone(input, with_timezone=False)
+        assert expected == output_2
 
 
-@pytest.mark.parametrize("input, expected", TEST_UTC_TIMEZONE)
-def test_utc_timezone(input, expected):
-    output_1 = ensure_utc_timezone(input)
-    assert expected == output_1
-    output_2 = add_or_remove_timezone(input, with_timezone=True)
-    assert expected == output_2
+@pytest.mark.parametrize("timezone, input, expected", TEST_UTC_TIMEZONE)
+def test_utc_timezone(timezone, input, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output_1 = ensure_utc_timezone(input)
+        assert expected == output_1
+        output_2 = add_or_remove_timezone(input, with_timezone=True)
+        assert expected == output_2
 
 
-def test_get_now_datetime():
-    output_1 = get_now_datetime(with_timezone=False)
-    assert output_1.tzinfo is None
-    output_2 = get_now_datetime(with_timezone=True)
-    assert output_2.tzinfo is not None
-    assert output_2.tzinfo == UTC
+# @pytest.mark.parametrize("timezone", TIMEZONES)
+# Due to a bug in freezegun (https://github.com/spulec/freezegun/issues/348, https://github.com/spulec/freezegun/issues/553)
+# this only works with timezone = UTC
+@pytest.mark.parametrize("timezone", [datetime.timedelta(hours=0)])
+def test_get_now_datetime_w_timezone(timezone):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output_2 = get_now_datetime(with_timezone=True)
+        assert output_2.tzinfo is not None
+        assert output_2.tzinfo == UTC
+        assert output_2 == datetime.datetime(2024, 2, 3, 4, 5, 6, tzinfo=UTC)
 
 
-@pytest.mark.parametrize("seconds, timestamp", TEST_EPOCH_SECONDS)
-def test_epoch_seconds(seconds, timestamp):
-    ts_wo_tz = datetime.datetime(**timestamp)
-    assert seconds == get_epoch_seconds(ts_wo_tz)
-    timestamp_w_tz = dict(timestamp)
-    timestamp_w_tz['tzinfo'] = UTC
-    ts_w_tz = datetime.datetime(**timestamp_w_tz)
-    assert seconds == get_epoch_seconds(ts_w_tz)
-    output_1 = from_epoch_seconds(seconds, with_timezone=False)
-    assert ts_wo_tz == output_1
-    output_2 = from_epoch_seconds(seconds, with_timezone=True)
-    assert ts_w_tz == output_2
+@pytest.mark.parametrize("timezone", TIMEZONES)
+def test_get_now_datetime_wo_timezone(timezone):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output_1 = get_now_datetime(with_timezone=False)
+        assert output_1.tzinfo is None
+        assert output_1 == datetime.datetime(2024, 2, 3, 4, 5, 6)
 
 
-@pytest.mark.parametrize("timestamp, expected_seconds", TEST_EPOCH_TO_SECONDS)
-def test_epoch_to_seconds(timestamp, expected_seconds):
-    assert expected_seconds == get_epoch_seconds(timestamp)
+@pytest.mark.parametrize("timezone, seconds, timestamp", TEST_EPOCH_SECONDS)
+def test_epoch_seconds(timezone, seconds, timestamp):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        ts_wo_tz = datetime.datetime(**timestamp)
+        assert seconds == get_epoch_seconds(ts_wo_tz)
+        timestamp_w_tz = dict(timestamp)
+        timestamp_w_tz['tzinfo'] = UTC
+        ts_w_tz = datetime.datetime(**timestamp_w_tz)
+        assert seconds == get_epoch_seconds(ts_w_tz)
+        output_1 = from_epoch_seconds(seconds, with_timezone=False)
+        assert ts_wo_tz == output_1
+        output_2 = from_epoch_seconds(seconds, with_timezone=True)
+        assert ts_w_tz == output_2
 
 
-@pytest.mark.parametrize("relative_time_string, with_timezone, now, expected", TEST_CONVERT_RELATIVE_TO_DATETIME)
-def test_convert_relative_to_datetime(relative_time_string, with_timezone, now, expected):
-    output = convert_relative_to_datetime(relative_time_string, with_timezone=with_timezone, now=now)
-    assert expected == output
+@pytest.mark.parametrize("timezone, timestamp, expected_seconds", TEST_EPOCH_TO_SECONDS)
+def test_epoch_to_seconds(timezone, timestamp, expected_seconds):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        assert expected_seconds == get_epoch_seconds(timestamp)
 
 
-@pytest.mark.parametrize("input_string, input_name, backend, with_timezone, now, expected", TEST_GET_RELATIVE_TIME_OPTION)
-def test_get_relative_time_option(input_string, input_name, backend, with_timezone, now, expected):
-    output = get_relative_time_option(input_string, input_name, backend=backend, with_timezone=with_timezone, now=now)
-    assert expected == output
+@pytest.mark.parametrize("timezone, relative_time_string, with_timezone, now, expected", TEST_CONVERT_RELATIVE_TO_DATETIME)
+def test_convert_relative_to_datetime(timezone, relative_time_string, with_timezone, now, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output = convert_relative_to_datetime(relative_time_string, with_timezone=with_timezone, now=now)
+        assert expected == output
+
+
+@pytest.mark.parametrize("timezone, input_string, input_name, backend, with_timezone, now, expected", TEST_GET_RELATIVE_TIME_OPTION)
+def test_get_relative_time_option(timezone, input_string, input_name, backend, with_timezone, now, expected):
+    with freeze_time("2024-02-03 04:05:06", tz_offset=timezone):
+        output = get_relative_time_option(input_string, input_name, backend=backend, with_timezone=with_timezone, now=now)
+        assert expected == output

--- a/tests/unit/requirements.txt
+++ b/tests/unit/requirements.txt
@@ -7,5 +7,6 @@ cryptography
 idna
 ipaddress ; python_version < '3.0'
 
+freezegun
 unittest2 ; python_version < '2.7'
 importlib ; python_version < '2.7'

--- a/tests/unit/requirements.txt
+++ b/tests/unit/requirements.txt
@@ -7,6 +7,7 @@ cryptography
 idna
 ipaddress ; python_version < '3.0'
 
-freezegun
+freezegun == 0.3.10 ; python_version < '2.7'
+freezegun ; python_version >= '2.7'
 unittest2 ; python_version < '2.7'
 importlib ; python_version < '2.7'

